batchproducer/batchproducer.go

```diff
@@ -223,15 +235,44 @@ func (b *batchProducer) run() {
 	}
 }
 
-// from/for interface BatchProducer
+// from/for interface Producer
 func (b *batchProducer) Stop() error {
-	// TODO: Immediately stop accepting new records by Add, then block until all the records in the buffer have been sent
 	if b.isRunning() {
 		b.stop <- true
 	}
 	return nil
 }
 
+// from/for interface Producer
+// TODO: send all batches in parallel, will require broader refactoring
+func (b *batchProducer) Flush(timeout time.Duration) (int, error) {
+	b.Stop()
+
+	timer := time.NewTimer(timeout)
+	if timeout == 0 {
+		timer.Stop()
+	}
+
+	timedOut := false
+
+loop:
+	for len(b.records) > 0 {
+		select {
+		case <-timer.C:
+			timedOut = true
+			break loop
```

Comment:
Oh wow, ok. I misunderstood your first comment. I was assuming the `break loop` was a Go equivalent of a `goto loop`. Understood now :+1: